drivers/base/bootvid/common.c

```diff
@@ -82,24 +82,28 @@ BitBlt(
     PrepareForSetPixel();
 
     /* 4bpp blitting */
-    for (dy = Top; dy < Bottom; ++dy)
+    for (Y = Top; Y < Bottom; Y++)
     {
-        sx = 0;
-        do
-        {
-            /* Extract color */
-            color = Buffer[offset + sx];
+        InputBuffer = Buffer;
 
-            /* Calc destination x */
-            dx = Left + (sx << 1);
+        for (X = Left, Pixel = 0;
+             X < Right;
+             ++X, ++Pixel)
+        {
+            if (Pixel % 2)
+            {
+                SetPixel(X, Y, Colors & 0x0F);
+            }
+            else
+            {
+                /* Extract colors at every two pixels */
+                Colors = *InputBuffer++;
```

Comment:
Suggestion (so that we get to see first the first case):
```c
            if ((Pixel % 2) == 0)
            {
                /* Extract colors for two pixels at once */
                Colors = *InputBuffer++;
                /* Set the first pixel... */
                SetPixel(X, Y, Colors >> 4);
            }
            else
            {
                /* ... then the second pixel */
                SetPixel(X, Y, Colors & 0x0F);
            }
```